### ISSUE TITLE
Next.js components from `"next"` package are not importing types properly

### ISSUE THREAD

**Issue by richardvanbergen** [denoland/deno#25359]
I was excited to try out deno on Next.JS after hearing about it on the syntax podcast. It runs well but unfortunately I'm getting a couple of type errors when importing the next core components like `next/link` and `next/image`.

This only happens in the editor with the LSP so it might not be the right place to post it so let me know.

For good measure in the project that I found it in I tried copy and pasting the compiler options into `deno.json`. It had no effect.

> JSX element type 'Link' does not have any construct or call signatures.deno-ts(2604)
> 'Link' cannot be used as a JSX component.
  Its type 'typeof import("file:///Users/richardvanbergen/code/deno-nextjs-bug-example/node_modules/.deno/next@14.2.7/node_modules/next/link")' is not a valid JSX element type.deno-ts(2786)

> JSX element type 'Image' does not have any construct or call signatures.deno-ts(2604)
'Image' cannot be used as a JSX component.
>  Its type 'typeof import("file:///Users/richardvanbergen/code/deno-nextjs-bug-example/node_modules/.deno/next@14.2.7/node_modules/next/image")' is not a valid JSX element type.deno-ts(2786)

Reproduction Steps:

1. Download the repo from the link below
2. Enable Deno from the menu command bar `Deno: Enable`
3. Ensure Deno Future is checked in the user settings
4. Just for good measure, add `DENO_FUTURE=1` explicitly in the settings.
5. Navigate to `src/app/page.tsx`.
6. Behold! Errors!

Version: Deno 1.46.2
Reproduction repo: [https://github.com/richardvanbergen/deno-nextjs-bug-example](https://github.com/richardvanbergen/deno-nextjs-bug-example)

Side note: I can deal with it but is there a way to enable wildcard import aliases?
Side side note: There's also an error importing the styles I personally don't care about that but if it's easy to answer then it might be worth documenting it.

**Comment by nayeemrmn** [comment]
This has the same cause as #24451. It would work if you replace `<Link>` with `<Link.default>`, but that shouldn't be necessary. It's  a quirk of tsc's esm interop that isn't integrated here properly.

**Comment by richardvanbergen** [comment]
Thanks for looking at this so quickly.

For anyone else searching the workaround I'm now using is.

```ts
// rexport next default exports due to this bug: https://github.com/denoland/deno/pull/25484
// remove this when it's fixed

import NextLink from 'next/link'
import NextImage from 'next/image'

const Link = NextLink.default
const Image = NextImage.default

export { Link, Image }
```

```json
{
  "imports": {
    "@next": "./src/components/next.ts"
  }
}
```

```ts
import { Link } from '@next'
```

**Comment by nayeemrmn** [comment]
https://github.com/denoland/deno/pull/25484#issuecomment-2335046615

It turns out this isn't a bug for Deno. It's caused by https://github.com/vercel/next.js/issues/46078.

For Node, if you add `{ "type": "module" }` to your package.json and modify your tsconfig to the following Deno-like one:
```json
{
  "compilerOptions": {
    "allowJs": true,
    "esModuleInterop": true,
    "experimentalDecorators": false,
    "isolatedModules": true,
    "lib": ["esnext"],
    "module": "nodenext",
    "moduleDetection": "force",
    "noEmit": true,
    "resolveJsonModule": true,
    "strict": true,
    "target": "esnext",
    "useDefineForClassFields": true,
    "jsx": "react",
    "jsxFactory": "React.createElement",
    "jsxFragmentFactory": "React.Fragment",
    "types": ["@types/node"],
  },
}
```
You'll see the same error and it will necessitate rewriting `src/app/page.tsx` to this either way:
```tsx
import NextLink from "next/link.js";
import NextImage from "next/image.js";
import * as React from 'react';

const Link = NextLink.default;
const Image = NextImage.default;

export default function Home() {
  return (
    <>
      <Link href="..." />
      <Image src="..." alt="..." />
    </>
  );
}
```
Which would work with both tsc and Deno. I'll close this issue for now @richardvanbergen If you want please try out this config and toggle `deno.enable` off and on to confirm.

**Comment by yanickrochon** [comment]
How is this closed and marked completed?

**Comment by yanickrochon** [comment]
This Next.js also fails:
```ts
import localFont from "next/font/local";

const geistSans = localFont({
  src: "./fonts/GeistVF.woff",
  variable: "--font-geist-sans",
  weight: "100 900",
});
```
Saying that 
> This expression is not callable.
> Type 'typeof import("file:///home/yanick/dev/deno/variant/node_modules/next/font/local/index")' has no call signatures.deno-ts(2349)

However, the `next/font/local` has this single line:
```ts
export { default } from 'next/dist/compiled/@next/font/dist/local'
```
which, in turn, exposes:
```ts
export default function localFont<T extends CssVariable | undefined = undefined>(options: LocalFont<T>): T extends undefined ? NextFont : NextFontWithVariable;
```
and this is a perfectly valid function declaration.

Both `localFont` and `Link`, as the OP reported, evaluate just fine. No errors when the code is being executed. This is strictly a linting error, and it is a false one.

**Comment by karlhorky** [comment]
Yeah, there are quite a few issues with ESM fully-specified imports in Node.js, some of the latest: 

- https://github.com/vercel/next.js/issues/46078
- attempted PR by @balazsorban44 (Next.js team member)  https://github.com/vercel/next.js/pull/61525
- https://github.com/vercel/next.js/issues/47552
- https://github.com/vercel/next.js/pull/50357

Are The Types Wrong? and Publint also report errors:

- https://github.com/vercel/next.js/issues/54777#issuecomment-1817822937

**Comment by dmint789** [comment]
This bug gave me SUCH a headache to find. I was just trying to build my Next JS application, and I kept getting the error `Cannot read properties of undefined (reading 'bold')`. NOTHING about that error message suggests that this has something to do with the Link component, lol. Should I report this in the Next JS repo, so they can set up an error message for these cases?

In any case, is this bug (the one in this issue) more relevant to Deno or Next JS?

**Comment by dmint789** [comment]
I also just realized that the same error happens when I try to build, when there is a component somewhere in my project that imports a component that doesn't exist. It just gives me that cryptic error message. The same message also occurs when I use an incorrect component attribute (e.g. `<Link to=.../>` instead of `<Link href=.../>`).

**Comment by dmint789** [comment]
[This issue](https://github.com/denoland/deno/issues/21269) also displayed the same error message for a totally different bug.

**Comment by telmotrooper** [comment]
One would expect that with @ry himself [stating that we can use Next.js with Deno](https://youtu.be/swXWUfufu2w?t=119) that Deno's language server would behave better when you implement a simple Next.js web app following the [official documentation](https://docs.deno.com/runtime/tutorials/how_to_with_npm/next/).

My experience refers specifically to VS Code, with the [official Deno extension](https://marketplace.visualstudio.com/items?itemName=denoland.vscode-deno) and `"deno.enable": true` in `.vscode/settings.json`, as well as with [recommended JSX settings](https://docs.deno.com/runtime/reference/jsx/#jsx-automatic-runtime-(recommended)).

--

Just to be clear, I don't mean to be rude at all. Love the work you guys put on Deno!

**Comment by fengkiej** [comment]
```.vscode/settings.json
{
  "deno.enable": true,
  "deno.disablePaths": [
    "./src/pages",
    "./src/app",
    "./src/components",
  ]
}
```

**Comment by wanderer** [comment]
> This bug gave me SUCH a headache to find. I was just trying to build my Next JS application, and I kept getting the error `Cannot read properties of undefined (reading 'bold')`. NOTHING about that error message suggests that this has something to do with the Link component, lol. Should I report this in the Next JS repo, so they can set up an error message for these cases?
> 
> In any case, is this bug (the one in this issue) more relevant to Deno or Next JS?
> 
> Edit: I tried the workaround using `Link.default`, but I still get the same error. If someone needs a reproducible example, I can make a repo.

you need to do `deno run --unstable-unsafe-proto -A npm:next build` the `Cannot read properties of undefined (reading 'bold')` is coming from [chalk](https://github.com/denoland/deno/issues/21269)

**Comment by dmint789** [comment]
> > This bug gave me SUCH a headache to find. I was just trying to build my Next JS application, and I kept getting the error `Cannot read properties of undefined (reading 'bold')`. NOTHING about that error message suggests that this has something to do with the Link component, lol. Should I report this in the Next JS repo, so they can set up an error message for these cases?
> 
> you need to do `deno run --unstable-unsafe-proto -A npm:next build` the `Cannot read properties of undefined (reading 'bold')` is coming from [chalk](https://github.com/denoland/deno/issues/21269)

Oh, that worked! Why is this not the default behavior? Also, why is the flag called "unstable"? I thought the convention was to call those kinds of flags "experimental".

**Comment by dmint789** [comment]
Turns out, this is not only a problem with Next JS components. For example, the import from the `jwt-decode` library also errors out saying `This expression is not callable.`

**Comment by veryhealthy** [comment]
Hello everyone,

I initiated a next js project with app router and came to the same type error.

```ts
import Image from "next/image";
'Image' cannot be used as a JSX component.
```

Replacing `<Image> by <Image.default>` is working.

Thank you for good work.

**Comment by fro-profesional** [comment]
> This bug gave me SUCH a headache to find. I was just trying to build my Next JS application, and I kept getting the error `Cannot read properties of undefined (reading 'bold')`. NOTHING about that error message suggests that this has something to do with the Link component, lol. Should I report this in the Next JS repo, so they can set up an error message for these cases?
> 
> 
> 
> In any case, is this bug (the one in this issue) more relevant to Deno or Next JS?
> 
> 
> 
> Edit: I tried the workaround using `Link.default`, but I still get the same error. If someone needs a reproducible example, I can make a repo.

I had the exact same issue, for me it was that I had deno code using jsr and nextjs normal app, I had to exclude in tsconfig the path for the deno script, so the nextjs doesn't pick it

**Comment by stefvw93** [comment]
When you are creating a project with a tsconfig.json file (like NextJS has), it is recommended by the IDE plugin to disable the LSP for that particular project.

For example, I am using NextJS in a deno workspaces environment, and I have to disable the LSP for the NextJS app to get proper type support.

I am using VS Code, and the Deno plugin reports:

> A tsconfig.json with compiler options was discovered in a Deno-enabled folder. For projects with compiler plugins, it is recommended to disable the Deno language server if you are seeing errors.

I have a monorepo structure like:
```
apps/web (nextjs)
packages/ui
packages/lib
```

I updated my deno workspace settings in .vscode/settings.json like so:

```json
{
  "deno.enable": true,
  "deno.disablePaths": [
    "apps/web"
  ]
}
```

With these settings, VS Code will use the default Typescript LSP, rather than the Deno LSP for my NextJS app.

I agree it does not feel ideal, but understandable given the fact that NextJS provides its own build tooling and Typescript configuration.

**Comment by franckdervaux** [comment]
None of these worked for me. I am not using the Image or Link component, nor do I use chalk. But, the issue appeared as soon as I added prima to the demo nextjs project. The issue does not seem to be related to VSCode since I am getting the same error when running `demo task build` in a separate terminal. Anyone could help me out?

**Comment by Sc1entifik** [comment]
> Hello everyone,
> 
> I initiated a next js project with app router and came to the same type error.
> 
> ```ts
> import Image from "next/image";
> 'Image' cannot be used as a JSX component.
> ```
> 
> Replacing `<Image> by <Image.default>` is working.
> 
> Thank you for good work.

I tried changing the import to ``` import Image.default from "next/image"; ``` and also the tag to ``` <Image.default /> ``` and tried both together to no avail.

**Comment by awkaiser** [comment]
# 🙈🙉🙊 Deno 2.1.4 + Next.js 15.1.2

Referencing `[Component].default` can cause other issues:

> Error: Cannot access Image.default on the server. You cannot dot into a client module from a server component. You can only pass the imported name through.

Littering your JSX with [`@ts-expect-error`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#-ts-expect-error-comments) is tedious but will:

1. Silence error when using the Deno LSP

    a. ‼️ Any _other_ TS errors associated with JSX line where component is used will _also_ be disregarded

2. Throw an `Unused '@ts-expect-error' directive` error when problem is resolved

```tsx
return (
  <main>
    {/* @ts-expect-error */}
    <Image src='/next.svg' />
  </main>
);
```

**Comment by dmint789** [comment]
I'm a bit confused about this one. Why is this issue with .default so hard to fix? I get this with other libraries too, not just Next JS. Is it something that can't be solved unless the library maintainers fix it on their end?

**Comment by awkaiser** [comment]
# 🫡 `[welcometothepartypal.gif]`

> Is it something that can't be solved unless the library maintainers fix it on their end?

Correct, it's not a problem for Deno to solve.

While having `npm:*` interoperability is convenient and will boost Deno adoption, it also reopens its community to the pain of NPM packages missing accurate **\*.d.ts** files and/or not having the **package.json** structure and published files necessary to work seamlessly with ESM and CJS.[^1] This isn't Deno specific pain, it also affects Node projects using [`type: "module"`](https://nodejs.org/docs/latest-v22.x/api/packages.html#type) + [`moduleResolution: "nodenext"`.](https://www.typescriptlang.org/tsconfig/#moduleResolution) Getting this right is tricky, and nobody can say that the move from CJS to ESM has been easy smooth sailing for the Node / NPM community.

----

This issue is helpful as a paper trail for folks to follow as they inevitably trip over type errors RE: promises of "Deno can run Next.js now!" but [the earlier choice to close](https://github.com/denoland/deno/issues/25359#event-14168477749) was, IMHO, technically correct.

I understand the desire to pursue "pure Deno development" with its LSP but, as of this writing, people should follow the Deno team's recommendation to disable the Deno LSP on projects with **tsconfig.json** compiler options unsupported by Deno. Their [VSCode extension](https://docs.deno.com/runtime/reference/vscode/) helpfully [suggests](https://github.com/denoland/vscode_deno/pull/1166) this!

[^1]: ["[NEXT-779] next/* - Typescript cannot find module when moduleResolution=nodenext and type=module"](https://github.com/vercel/next.js/issues/46078#issuecomment-2134508560)

**Comment by dmint789** [comment]
@awkaiser I had a look at the Next JS repo, and wasn't able to find an issue related to this. Since I'm not really a Typescript/ESM expert, I don't feel qualified to open this particular issue. Do you think you could open it on their end so they can take care of it so we can use the Deno LSP for Next JS projects in the future? I personally really don't like the option of disabling the Deno LSP for my projects, after putting in all of that effort migrating to Deno.

**Comment by awkaiser** [comment]
@dmint789 I see no reason to open a new issue. There are already [open issues](https://github.com/denoland/deno/issues/25359#issuecomment-2419641885) such as [**NEXT-779**](https://github.com/vercel/next.js/issues/46078) that, while they may not discuss Deno specifically, will also fix the problem for Deno when resolved. 🍻

**Comment by richardvanbergen** [comment]
> This issue is helpful as a paper trail for folks to follow as they inevitably trip over type errors RE: promises of "Deno can run Next.js now!" but [the earlier choice to close](https://github.com/denoland/deno/issues/25359#event-14168477749) was, IMHO, technically correct.

I think for me the main issue I had wasn't that it doesn't work with next but that this only seemed to be a problem for the LSP but not the compiler. There might be a reason for that (Runtimes and LSPs are not my thing) but it is unintuitive behaviour. The compiler ran the program just fine.

**Comment by dimikot** [comment]
> That said, if you're not just kicking Deno's tires and are planning to build a serious Next.js project with it

That's a super-true and bold analogy about kicking the tires, ha-ha. Respect.

I found one work-around to not use `@ts-expect-error` - create a folder `src/deno-quirks` (BTW maybe not deno-quirks but modern-esm-quirks since it's not quite Deno who's mostly guilty but Next) and put a file there:

```ts
import NextLink from "next/link";

export const Link = NextLink as unknown as typeof NextLink.default;
```

Then everywhere you use `<Link/>` (or any other similar default-exports), import it from that folder.

**Comment by RaiPeter** [comment]
> Thanks for looking at this so quickly.
> 
> For anyone else searching the workaround I'm now using is.
> 
> // rexport next default exports due to this bug: https://github.com/denoland/deno/pull/25484
> // remove this when it's fixed
> 
> import NextLink from 'next/link'
> import NextImage from 'next/image'
> 
> const Link = NextLink.default
> const Image = NextImage.default
> 
> export { Link, Image }
> {
>   "imports": {
>     "@next": "./src/components/next.ts"
>   }
> }
> import { Link } from '@next'

this was helpful as i was also trying out deno with nextjs dinosaur app and I got stuck with this issue for an hour.

**Comment by mintydev789** [comment]
@nayeemrmn could we please at least get an `--unstable-...` flag to have this issue resolved? The issue happens with multiple packages, not just Next JS. I've used Deno ever since the 2.0 release, and this single issue has been causing me the most issues. I (and many of us in the Deno community) really just want to be able to use the Deno LSP while developing. Could the LSP just somehow check if it can get `.default` when it gets the `deno-ts(2604)` error?

**Comment by dsherret** [comment]
I opened https://github.com/denoland/deno/issues/29673 to support `"moduleResolution": "bundler"` for type checking, but this would not be an issue if the "next" package didn't incorrectly describe itself as having a default export in its types when it really has an `export =` export. I'm going to close this issue because it doesn't seem to be an issue in Deno as previously pointed out. Follow https://github.com/vercel/next.js/issues/46078 for updates.

**Comment by nayeemrmn** [comment]
Reopening for tracking since we have a fix: #30603.

**Comment by mintydev789** [comment]
@nayeemrmn thank you for fixing this! Does this mean that Next JS components should now work if you have this in your `tsconfig.json`?

```json
"compilerOptions": {
    ...
    "moduleResolution": "bundler",
    ...
}
```

**Comment by nayeemrmn** [comment]
> Does this mean that Next JS components should now work if you have this in your `tsconfig.json`?
> 
> "compilerOptions": {
>     ...
>     "moduleResolution": "bundler",
>     ...
> }

@mintydev789 Yes, it should generally already be there with `npx create-next-app`.

We now expect it to work in the same cases as Node with `{ "type": "module" }`, in fact it would break the same way with Node in that situation if you changed it to `"moduleResolution": "nodenext"` etc.: https://github.com/vercel/next.js/issues/46078

**Comment by mintydev789** [comment]
@nayeemrmn I upgraded to Deno 2.5.4, but the issue is still not resolved. I still see the following errors:

```
JSX element type does not have any construct or call signatures.
Link cannot be used as a JSX component.
```

**Comment by nayeemrmn** [comment]
@mintydev789 Are you using `"moduleResolution": "bundler"`? Please provide a reproduction

**Comment by mintydev789** [comment]
> @mintydev789 Are you using `"moduleResolution": "bundler"`? Please provide a reproduction

Yes, I am. Well, I guess my reproduction is just [my repo](https://github.com/cubing/cubingcontests.com/tree/auth) (that specific branch). My Next JS app is under `client/`, and the types for `<Link/>` are not working, it's highlighted as an error still.

**Comment by mintydev789** [comment]
> @mintydev789 Are you using `"moduleResolution": "bundler"`? Please provide a reproduction

Kind ping :)

**Comment by nayeemrmn** [comment]
@mintydev789 You have to delete the `compilerOptions` in `client/deno.jsonc`, otherwise it overrides all options specified in `client/tsconfig.json`, including `"moduleResolution": "bundler"`.

It's mentioned in the third bullet point in https://docs.deno.com/runtime/reference/ts_config_migration/#tsconfig.json-compatibility.

**Comment by mintydev789** [comment]
@nayeemrmn ohh, I see now. Thanks a lot!
